### ISSUE TITLE
add an api to add new functiosn during stable releases.

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#4751]
This is basicially a spit of from https://github.com/wesnoth/wesnoth/pull/3576 because there   i didnt get that much feedack on idea.

Oftentimes one would like to add a new api function to wesnoth, in the following example assume we'd like to add a new function `wesnoth.set_dialog_tooltip` to set a dialogs widgets tooltip.

First lets recall what is as far as i know the main reason why we currently don't put new api functions in stable releases: we want addons to also run in slightly outdated versions, that is, we don't add for example 1.16.7 api, because then people would write addons that uses this api, and then would no longer work on 1.16.6 and older. 

Now, i really think the current approach of forbidding all new api functions is too restrictive, in particular in a case of our example function `set_dialog_tooltip` that causes little harm if omitted. Lets assume we have decided to add the function `set_dialog_tooltip` in 1.16.7 under 1.16.6 the following code would  crash, because the function `wesnoth.set_dialog_tooltip` can not be found:

```
set_dialog_tooltip(str, patch, to, widget)
```

The following code does work and is perfectly safe without causing any harm on any version

```
if compare_version(wesnoth_version, >= "1.16.7") then
  set_dialog_tooltip(str, patch, to, widget)
end
```

So the proposal of allow adding enw api function in stable, but force people that want to use them to check the wesnoth version before, to do that i propose a function called `wesnoth.switch_version` (we can also choose a different name here), that work similar to a `if compare_version` above: it compares the versions and calls the first block if  wesnoth is new enough and the second function if not. The idea is now that function added for example after  1.16.7 may only be called within such a wesnoth.switch_version block
```
-- this works in 1.16.7 and 1.16.6
wesnoth.switch_version("1.16.7", 
  function()
    -- 1.16.7 or later
    set_dialog_tooltip(str, patch, to, widget)
  end,
  funcion()
    -- 1.16.6 or older
    -- do nothing.
  end)
```
```
-- calling the function without switch versions will fail on both 1.16.6 adn 1.16.7
set_dialog_tooltip(str, patch, to, widget)
```

**Comment by stevecotton** [comment]
Add-ons can already depend on other add-ons, I think a better way to do this would be to have the public API in a new type of "library add-on", suppose one called "lib_gfgtdf" that defines a `gfgtdf.set_dialog_tooltip` function. All the version-checking code would be in the library add-on, and it has to have the code gracefully handle old Wesnoth versions.

UMC would depend on the library add-on, and the UMC code wouldn't need any version check other than declaring the dependency. UMC code just calls `gfgtdf.set_dialog_tooltip(...)`.

The API between UMC and library would be stable, but there's no need for a stable API between the library and Wesnoth itself.

**Comment by gfgtdf** [comment]
functions like set_dialog_tooltip can only be implemented in C++. Also with your solution the umc devs woudl have to change tehir code again in next stable version (here: 1.18) when the function becomes a 'normal' function and gets renames from gfgtdf.set_dialog_tooltip to wesnoth.set_dialog_tooltip

**Comment by GregoryLundberg** [comment]
All I see is the need for a way to test the version.

Probably need to be able to do all relations ==, !=, <, <=, >, >=.

Both Lua and WML should be able to do the test.

At present we just error off (crash and burn?) if one calls an unknown function. Failure to perform the version test should do the same if the version does not have the function. The idea of the test is simply to avoid the crash. Seems simple enough and won't cause problems if the test is not removed after the next "stable" release.

The question is more, is the feature being added (not the version-test but the tool tip or whatever) really important enough to slip into later "stable" versions and not wait for the next. Or is this really just a work-around for slow release cycles, which might be better handled by improving the release timing?

**Comment by gfgtdf** [comment]
> At present we just error off (crash and burn?) if one calls an unknown function. Failure to perform the version test should do the same if the version does not have the function. The idea of the test is simply to avoid the crash. Seems simple enough and won't cause problems if the test is not removed after the next "stable" release.

Not sure how you mean this.


> The question is more, is the feature being added (not the version-test but the tool tip or whatever) really important enough to slip into later "stable" versions and not wait for the next. Or is this really just a work-around for slow release cycles, which might be better handled by improving the release timing?

Well first the tooltip function is just an example (right now we could just add it to 1.15 without problems, this is really for features that we might want after 1.16), but to answer your question, a bit of both, from my experience as a umc dev that also knows the c++ code its quite frustrating if a very basic function (usually something that is already implemented in the c++ code but lacks a lua/wml api) is missing. But also improving the release timing makes little sense when there are simply not that many new things to justify a new major release, also considering that a new release still implies compatibility breaks of multiple types (saves no longer working due to changed mainline campaigns wml code, addons no longer working due to compatibility breaking api, mp games cannot accept people from different version due to oos errors, old scenario might be unbalanced because of ai improvements and balancing changes), i also wouldn't say its clear that that we should have new 'stable release' more often.

To make my position clearer, a 'new stable release' with the current meaning of 'adding new player visible stuff' and 'compatibility breaking changes' can (because of the 'adding new player visible stuff' part) and should (because of the 'compatibility breaking' part) not happen too often. So i really wouldn't call this a work around for our rather slow release cycle because i don't think a faster release cycle is currently that desirable.

**Comment by stevecotton** [comment]
> functions like set_dialog_tooltip can only be implemented in C++.

The library add-on would be a shim. Inside the library `gfgtdf.set_dialog_tooltip` is going to be calling `wesnoth.set_dialog_tooltip`, but that's a detail for the library's author to work out.

> Also with your solution the umc devs woudl have to change tehir code again in next stable version (here: 1.18) when the function becomes a 'normal' function and gets renames from gfgtdf.set_dialog_tooltip to wesnoth.set_dialog_tooltip

No they wouldn't. I'm assuming that the library add-on will also be available on 1.18, so that UMC authors who are still supporting both 1.16 and 1.18 can use the same UMC code on both. Once they drop 1.16 support from their UMC then the UMC code can be changed to drop the dependency; but the library add-on should probably go through a deprecation cycle (supported on 1.16 and 1.18, but not ported to 1.20).

The other thing is that, from a UMC author's perspective, I'd expect the shim's API to be stable. If the following sequence of events happens, I expect it to be the responsibility of the library's author to provide compatibility with the released-on-1.14.x API of gfgtdf.set_dialog_tooltip.
* wesnoth.set_dialog_tooltip appears in the 1.15.3 release
* gfgtdf.set_dialog_tooltip is implemented based on the 1.15.3 version
* the shim library is released on 1.14.x
* sometime between 1.15.3 and 1.16.0, someone adds a new argument to wesnoth.set_dialog_tooltip
* gfgtdf.set_dialog_tooltip needs to continue being compatible with the already-written UMC code

**Comment by GregoryLundberg** [comment]
I don't get why you're worrying about the function. It seems immaterial to the argument.

What you want is a function/macro which is presently FALSE which can be made to become TRUE for some future release and all following releases. Actually you want a template (style, class, whatever) which provides for many such feature-tests.

You want it to test versions. But that's a special case of "feature test" ... the version number feature. You might also want the "macOS feature" or the "touch screen" feature or some other feature.

Who cares if the UMC has a no-longer-needed feature test? Maybe, if there's a few thousand of them, you'd want to rewrite to drop them. But leaving them in forever means that a newer UMC can be written to gracefully fall back to an ancient version, if one is so inclined.

**Comment by gfgtdf** [comment]
@stevecotton i don' see how your proposal woudl fix any of the things that this one is supposed to give:
- if the library addon calls `wesnoth.set_dialog_tooltip` we need to make a way to make a way to make sure its only available on wesnoth versions that have that function available.
- making an addon dependent on this function would then make this addon reauite the newer wesnoth version which is exactly want we don't want in our example above.

furthermore even with deprecation time for the alias your poposal still seems to be simply more work for both the umc dev who, even ( be it after the deprecation phase) will need to change the code eventually, And for us who needs to maintain an addon.





> What you want is a function/macro which is presently FALSE which can be made to become TRUE for some future release and all following releases. Actually you want a template (style, class, whatever) which provides for many such feature-tests.
> You want it to test versions. But that's a special case of "feature test" ... the version number feature. You might also want the "macOS feature" or the "touch screen" feature or some other feature.

No i only want versions for now. More importantly: the whole point of this fr is that i want to **force** people to check for the existence of the feature before using it, by providing an api that only enables a feature when people checked for it (on our example `set_dialog_tooltip` will only work if its called from within the `switch_version` function). Just checking for a version is something that can already be done with `wesnoth.compare_version`.


> Who cares if the UMC has a no-longer-needed feature test? Maybe, if there's a few thousand of them, you'd want to rewrite to drop them. But leaving them in forever means that a newer UMC can be written to gracefully fall back to an ancient version, if one is so inclined.

i have no idea to what you are refering here.

**Comment by GregoryLundberg** [comment]
The reason you don't know what I'm talking about is you're forcing the feature to be disabled unless guarded by a feature-test. I assume that people will try to use it without the test and it will fail in some way.

If you're forcing a feature-test guard, I don't see why you would ever want to allow the function in an un-guarded context. So there should never be a need to change it to be "always available" as discussed above. Otherwise you'll get UMC authors complaining it's too confusing. When should I use the guard and old name, and when should I use the new name? Why can't we just have one rule?

**Comment by stevecotton** [comment]
> if the library addon calls wesnoth.set_dialog_tooltip we need to make a way to make a way to make sure its only available on wesnoth versions that have that function available.

Other way round - the library add-on has to check which version of Wesnoth it's running on, and only call the function if it's available. That responsibility includes handling multiplayer where some players have different versions.

> furthermore even with deprecation time for the alias your poposal still seems to be simply more work for both the umc dev who, even ( be it after the deprecation phase) will need to change the code eventually, And for us who needs to maintain an addon.

What situations would need effort to maintain the add-on? I agree that there are situations that would need it, but I think the same situations with your proposal would make every UMC author have to spend the same effort on their addon.

**Comment by gfgtdf** [comment]
> Other way round - the library add-on has to check which version of Wesnoth it's running on, and only call the function if it's available. That responsibility includes handling multiplayer where some players have different versions.

okay i see, this would work in the case of the `set_tooltip` function mentioned above, when its clear what to do when the function is unavailable (doing nothing). Buti think  there might be functions where there what do do alterntiveley really depends on the calling concreate usecase.

Anyways i think your idea might work for this case, but i'd still reccomend to name the function provided by the addon wesnoth.set_tooltip instead of gfg.set_tooltip so that people don't have to rename if they want to adapt for the nest stable.

Also if we want to do this, we have to make sure 'addon dependency' works properly, as there might be quirks relatred to it in particular wrt to multiplayer, sine the mp 'reauire addon' and the 'addon dependency' are implemented independently of each other iirc.

> I assume that people will try to use it without the test and it will fail in some way.

Yes this exactly  what i want. (to be clear, 'disabling the feature' here means behaving like the older version, that is, failing  with a function 'name' not found error, )


> When should I use the guard and old name, and when should I use the new name? Why can't we just have one rule?

I never proposed different names, please quote exactly what you are referring  to .If you are referring to stevecottons suggestions then make this clear by quoting stevecotton before your comments.


> ETA: Oh, and you may only want the feature-test for this one case but it's still just a case of a larger issue

Yes maybe, but i'd really like to solve the discussion on this features design first, and then afterwards we can discuss whether we should make it more generic to also extend to other features.

**Comment by stevecotton** [comment]
> i'd really like to solve the discussion on this features design first, and then afterwards we can discuss whether we should make it more generic to also extend to other features.

I don't think we can have a full discussion of `wesnoth. switch_version` using only an example where the "else" block is simply `-- do nothing`. That's too trivial to explore the edge cases.

**Comment by gfgtdf** [comment]
> I don't think we can have a full discussion of `wesnoth. switch_version` using only an example where the "else" block is simply `-- do nothing`. That's too trivial to explore the edge cases

By 'this feature' i didn't mean the toolip example, but rather the feature to 'backport' new functions, ( instead of plattform-dependent features that he brought up)

**Comment by GregoryLundberg** [comment]
You're correct, I was coming from the point-of-view that changing the name from version to version is a profoundly bad idea and only leads to confusion for the UMC author.

And I'm also coming from the point-of-view that if you're going to do this work you want to provide a means, going forward, for a feature to be adding in, say, 1.21.5 and 1.22-dev and not have to worry about what happens when we go to 1.23 or later.

But, I'm out. I feel I've achieved my purpose by raising some issues to think about. I know how I'd do it, and I'm sure that leaks into my discussion.

**Comment by stevecotton** [comment]
> By 'this feature' i didn't mean the toolip example, but rather the feature to 'backport' new functions, ( instead of plattform-dependent features that he brought up)

I don't understand this reply. By `wesnoth.switch_version` I meant the backporting feature, not the tooltip.

**Comment by gfgtdf** [comment]
> I don't understand this reply. By wesnoth.switch_version I meant the backporting feature, not the tooltip.

You replied on my comment where i said i didn't want to talk about feature tests for plattform dependent functions, by saying "using only an example where" which only makes sense if you misread my comment.

**Comment by stevecotton** [comment]
Ah, I understand the confusion now; I was replying to the "I'd really like to solve the discussion on this feature's design" without the other context. However, I don't think we can have a full discussion about the design of the backporting feature without considering a much more complex example.

Defining `wesnoth.switch_version` with two function arguments needs an example where both functions are non-trivial. I think this will lead to UMC having untested code paths, and OOS issues.
